src/eve/pattern_matching.py

```diff
@@ -0,0 +1,159 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import inspect
+from dataclasses import dataclass
+from functools import singledispatch
+from types import ModuleType
+from typing import Any, Iterator
+
+
+@dataclass(frozen=True)
+class ObjectPattern:
+    """Class to pattern match general objects.
+
+    A pattern matches an object if it is an instance of the specified
+    class and all attributes of the pattern (recursively) match the
+    objects attributes.
+
+    >>> class Foo:
+    ...    def __init__(self):
+    ...        self.bar = 1
+    ...        self.baz = 2
+    >>> assert ObjectPattern(Foo, {"bar": 1}).match(Foo())
```

Comment:
Slightly more realistic example:
```suggestion
    >>> class Foo:
    ...    def __init__(self, bar, baz):
    ...        self.bar = bar
    ...        self.baz = baz
    >>> assert ObjectPattern(Foo, {"bar": 1}).match(Foo(1, 2))
```